.github/workflows/main.yml

```diff
@@ -21,65 +24,72 @@ jobs:
     runs-on: ubuntu-latest
 
     steps:
-    - uses: actions/checkout@v2
+    - name: "Checkout repository"
+      id: checkout_repo
+      uses: actions/checkout@v3
+      with:
+        fetch-depth: 0
+        ref: ${{ github.event.pull_request.head.sha }}
 
-    - name: Lowercase repository name for docker build
+    - name: "Lowercase repository name for docker build"
       id: lowercase-repository-name
-      run: |
-        echo "HPE_BUILD_REPO=$(echo ${{ github.repository }} | tr '[:upper:]' '[:lower:]')" >> $GITHUB_ENV
-        echo "HPE_DEPLOY_REPO=$(echo ${{ github.repository }} | tr '[:upper:]' '[:lower:]')" >> $GITHUB_ENV
-
-    - name: set TAG & TESTTAG for main/master
-      if: ${{ github.event.repository.default_branch == 'main' || github.event.repository.default_branch == 'master' }}
-      run: |
-        echo "TAG=${{ github.sha }}" >> ${GITHUB_ENV}
-        echo "LATESTTAG=latest" >> ${GITHUB_ENV}
-        echo "TESTTAG=test-${{ github.sha }}" >> ${GITHUB_ENV}
+      run: echo "REPO_NAME=$(echo ${{ github.repository }} | tr '[:upper:]' '[:lower:]')" >> $GITHUB_ENV
 
-    - name: set TAG & TESTTAG for all other branches
-      if: ${{ github.event.repository.default_branch != 'main' && github.event.repository.default_branch != 'master' }}
+    - name: "Set tags for main/master"
+      id: set_tags
       run: |
-        echo "TAG=dev-${{ github.sha }}" >> ${GITHUB_ENV}
-        echo "LATESTTAG=dev-latest" >> ${GITHUB_ENV}
-        echo "TESTTAG=test-${{ github.sha }}" >> ${GITHUB_ENV}
-
-    - name: Docker login
+        echo "VERSION_TAG=$(./git-version-gen)" >> ${GITHUB_ENV}
+        echo "LATEST_TAG=latest" >> ${GITHUB_ENV}
+        echo "TEST_TAG=$(git rev-parse HEAD)-test" >> ${GITHUB_ENV}
+        echo "SHA_TAG=$(git rev-parse HEAD)" >> ${GITHUB_ENV}
+        echo "${GITHUB_ENV}:"
+        cat ${GITHUB_ENV}
+      shell: bash
+
+    - name: "Docker metadata"
+      id: meta
+      uses: docker/metadata-action@v4
+      with:
+        images: |
+          ghcr.io/${{ env.REPO_NAME }}
+        tags: |
+          # for merge to master branch, tag example: 'master'
+          type=ref,event=branch
+          # for PR event, tag example: 'pr-3'
+          type=ref,event=pr
+          # for PR event or merge event, tag example: 0.0.1.12-5667
+          type=raw,value=${{ env.VERSION_TAG }}
+          # for PR event or merge, tag example: 566769e04d2436cf5f42ae4f46092c7dff6e668e
+          type=raw,value=${{ env.SHA_TAG }}          
+          # for push to semver tag, tag example: 0.0.2
```

Comment:
We might want to update the comment to note that this will update the `latest` tag, I don't think any of these other cases do.